### PR TITLE
Swerve Drive Controller

### LINKED ISSUE #816: Rewrite igvc_platform motor_controller node to work with Swervi
To control the real world robot, we need to rewrite the motor controller node. This requires the firmware to be completed for Swervi, and will probably take a while to complete.

AC: Real life Swervi is able to be controlled with joystick, can navigate to waypoints.

### PR DESCRIPTION
**Pull request by VAM7686**
# Description

Swerve drive motor controller.

This PR does the following:
- Adds a new swerve drive motor controller node.
- Fixes issues with motor direction in the swervi urdf.
- Adds unit tests for the new swerve drive motor controller node.
- Fixes minor issue with gazebo swerve drive control node.
- Tunes PID controller for swivels

Fixes #816 

# Testing steps
Note: All movements should be in accordance with [REP 103](https://www.ros.org/reps/rep-0103.html).
## Test Case 1
1. Run `roslaunch igvc_gazebo ramp_lane.launch model:=1`
2. Run `roslaunch igvc_navigation swerve_drive.launch`
3. Run
```
rostopic pub /cmd_vel geometry_msgs/Twist "linear:
  x: 1.0
  y: 0.0
  z: 0.0
angular:
  x: 0.0
  y: 0.0
  z: 0.0"
```

Expectation: Swervi should go directly forward.

## Test Case 2
1. Run `roslaunch igvc_gazebo ramp_lane.launch model:=1`
2. Run `roslaunch igvc_navigation swerve_drive.launch`
3. Run
```
rostopic pub /cmd_vel geometry_msgs/Twist "linear:
  x: 0.0
  y: 1.0
  z: 0.0
angular:
  x: 0.0
  y: 0.0
  z: 0.0"
```

Expectation: Swervi should go to the left.

## Test Case 3
1. Run `roslaunch igvc_gazebo ramp_lane.launch model:=1`
2. Run `roslaunch igvc_navigation swerve_drive.launch`
3. Run
```
rostopic pub /cmd_vel geometry_msgs/Twist "linear:
  x: 0.0
  y: 0.0
  z: 0.0
angular:
  x: 0.0
  y: 0.0
  z: 1.0"
```

Expectation: Swervi should rotate counter clockwise.

## Test Case 4
1. Run `catkin_make run_tests_igvc_platform_rostest_src_tests_test_test_swerve_control.test`

Expectation: All tests pass.

# Self Checklist
- [x] I have formatted my code using `make format`
- [x] I have tested that the new behaviour works

### CODE REVIEW

File: igvc_navigation/src/swerve_drive/swerve_drive.cpp [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+  {
+    ROS_ERROR_STREAM("Unable to import parameters!");
+    return;
+  }
+  assertions::param(pNh, "max_vel", max_vel_, 3.0);
+  mbf_twist_ = nh.subscribe("/cmd_vel", 1, &SwerveDrive::twistToVelocity, this);
+  vel_pub_ = nh.advertise<igvc_msgs::velocity_quad>("/motors", 1);
+}
+
+void SwerveDrive::twistToVelocity(geometry_msgs::Twist twist)
```
Comment (a-stickan) [inline igvc_navigation/src/swerve_drive/swerve_drive.cpp:17]: Change to use const references

File: igvc_navigation/src/swerve_drive/swerve_drive.cpp [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+  vel_msg.bl_angle = wheel_info[1][1];
+  vel_msg.br_angle = wheel_info[2][1];
+  vel_msg.fr_angle = wheel_info[3][1];
+  vel_msg.duration = 0.02;
+  vel_msg.header.stamp = ros::Time::now();
+
+  vel_pub_.publish(vel_msg);
+}
+
+int SwerveDrive::set_command_angle(const double& target, const int wheel_idx)
```
Comment (a-stickan) [inline igvc_navigation/src/swerve_drive/swerve_drive.cpp:69]: Nit, but probably change `const int wheel_idx` to `const int& wheel_idx` for consistency

File: igvc_navigation/src/swerve_drive/swerve_drive.cpp [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+{
+  bool triple_point = false;
+  if (isclose(target, 0) || isclose(target, M_PI) || isclose(target, -M_PI))
+  {
+    triple_point = true;
+  }
```
Comment (a-stickan) [inline igvc_navigation/src/swerve_drive/swerve_drive.cpp:75]: Nit, but you could make this more elegant with a ternary operator deciding the assignment of `triple_point`

File: igvc_navigation/src/swerve_drive/swerve_drive.h [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  ros::NodeHandle nh;
+  ros::NodeHandle pNh;
+
+  ros::Subscriber mbf_twist_;
+  ros::Publisher vel_pub_;
+
+  std::array<std::array<double, 2>, 4> positions_list;
+  std::array<std::array<double, 2>, 4> limits_list;
+  std::array<std::array<double, 2>, 4> wheel_info;
+  double max_vel_{};
```
Comment (a-stickan) [inline igvc_navigation/src/swerve_drive/swerve_drive.h:27]: Is there a particular reason for zero initializing `max_vel_`?

**Review by a-stickan — CHANGES_REQUESTED** [review]
Looks good overall. All the test cases pass so the code is functional, but the const reference issues should be figured out so it can run as efficiently as possible.

File: igvc_navigation/src/swerve_drive/swerve_drive.h (hunk 4 above)
Comment (VAM7686) [reply igvc_navigation/src/swerve_drive/swerve_drive.h]: Not really. The assertions gives it a default value anyways. I'll just remove it.

**Review by a-stickan — APPROVED** [review]
Looks good!

**Pull request merged**